### ISSUE TITLE
Memory leak of ScanResults

### ISSUE THREAD

**Issue by jdeppe-pivotal** [classgraph/classgraph#233]
Hi, firstly thank you very much for this library - it is very useful!

I just updated our usage from 2.1.x to 4.0.4 and noticed a lot of out of memory issues with some of our test code. Digging in a bit, I noticed a lot of `ScanResult` objects retained on the heap. Indeed, running repeated `ClassGraph().scan()`s ends up chewing up a bunch of memory.

The culprit, I believe, is the shutdown hook being added in the `ScanResult` constructor. This ends up holding a reference to the `ScanResult` (and all it's objects) for the lifetime of the VM. I did a quick test of removing this bit of code and memory usage stayed constant regardless of how many iterations I ran.

It seems that this is the wrong place for this code. Seeing that you do provide a `close()` (cleanup) method, the user already has the ability to enable this functionality themselves - i.e. they could add the shutdownhook themselves if they wanted to or somehow schedule cleanup as relevant to the lifecycle of their app.

Thoughts?

**Comment by lukehutch** [comment]
@jdeppe-pivotal Great sleuthing, thanks for digging down to find the root of the problem.

This is tricky to solve. When you call close(), not only are some memory resources freed up, but temporary files are removed (such as inner jars-within-jars that had to be extracted in order to scan them). Those resources cannot be removed while the `ScanResult` is in scope, because calls like `ClassInfo#loadClass()` require those temporary files to be in place. And the `ClassInfo` objects obviously cannot be freed right after the scan, because they need to be returned by various `ScanResult` API calls.

There is already a method `ScanResult#finalize()` that calls `#close()`, and this is supposed to clean up resources on garbage collection (with the caveat that finalizers are of course not 100% reliable). I didn't think about the fact that the shutdown hook would hold a reference.

I changed the code to use `WeakReference<ScanResult>` in the shutdown hook. Can you please test the master version, and let me know if it solves the problem?

I'm also curious though -- how many scans are you performing? I'm surprised that the result of a scan caused a memory leak that was bad enough to even notice.

**Comment by lukehutch** [comment]
I also added information to the API documentation recommending that users call `ScanResult#close()` when they have finished with the `ScanResult`, to avoid unpredictably long GC pauses.

https://github.com/classgraph/classgraph/wiki/ClassGraph-API

**Comment by lukehutch** [comment]
Released in 4.0.5. That version should fix your issue, but please test and confirm.

**Comment by jdeppe-pivotal** [comment]
Thanks for your super-fast response! I'll test and get back to you.

**Comment by lukehutch** [comment]
Thanks! Please also see my question above about the number of scans you are performing.

**Comment by jdeppe-pivotal** [comment]
To answer your questions about the scans - there are only 4 happening with maybe 350-400 jars being scanned. It appears that the retained object size per ScanResult is about 45MB.

As to the fix - it turns out the anonymous inner class still ends up holding a reference to the enclosing class which still ends up causing a leak. One solution is to create a static helper class like this:
```
    ScanResult() {
...
        // Add runtime shutdown hook to remove temporary files on Ctrl-C or System.exit().
        // Uses a WeakReference so that garbage collection is not blocked. (#233)
        CleanupThread thread = new CleanupThread(new WeakReference<>(this));
        Runtime.getRuntime().addShutdownHook(thread);
    }

    public static class CleanupThread extends Thread {
        private final WeakReference<ScanResult> weakScanResult;

        public CleanupThread(WeakReference<ScanResult> weakScanResult) {
            this.weakScanResult = weakScanResult;
        }

        @Override
        public void run() {
            final ScanResult scanResult = weakScanResult.get();
            if (scanResult != null) {
                scanResult.close();
            }
        }
    }
```
If you could go to Java 8 syntax and lambdas, then this also works:
```
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            final ScanResult scanResult = scanResultRef.get();
            if (scanResult != null) {
                scanResult.close();
            }
        }));
```

**Comment by lukehutch** [comment]
Thanks, I haven't had to deal with weak references before, and forgot about the synthetic accessors holding references. It's also helpful to know that lambdas work more like anonymous static classes than anonymous inner classes.

I have checked in the change, and I'll push out another release shortly.

**Comment by lukehutch** [comment]
Released in 4.0.6. Thanks again for your guidance and help in getting this fixed!

**Comment by lukehutch** [comment]
@jdeppe-pivotal it looks like synthetic accessors are going away in JDK 11:

https://www.javacodegeeks.com/2018/08/nested-classes-private-methods.html
